diff --git a/tools/llvm-snippy/CMakeLists.txt b/tools/llvm-snippy/CMakeLists.txt
index 9911d613fca0..efdc58842434 100644
--- a/tools/llvm-snippy/CMakeLists.txt
+++ b/tools/llvm-snippy/CMakeLists.txt
@@ -123,4 +123,17 @@ endforeach()
 
 add_dependencies(llvm-snippy lld)
 
+install(TARGETS llvm-snippy
+  DESTINATION ${CMAKE_INSTALL_BINDIR}
+  COMPONENT llvm-snippy
+)
+install(FILES $<TARGET_FILE:lld>
+  DESTINATION ${CMAKE_INSTALL_BINDIR}
+  PERMISSIONS
+    OWNER_READ OWNER_WRITE OWNER_EXECUTE
+    GROUP_READ GROUP_EXECUTE
+    WORLD_READ WORLD_EXECUTE
+  RENAME ld.lld
+  COMPONENT llvm-snippy
+)
 
